go/types/opcache.go

```diff
@@ -8,15 +8,33 @@ import (
 	"encoding/binary"
 	"io/ioutil"
 	"os"
+	"sync"
+	"sync/atomic"
 
 	"github.com/attic-labs/noms/go/d"
 	"github.com/attic-labs/noms/go/hash"
 	"github.com/syndtr/goleveldb/leveldb"
 	"github.com/syndtr/goleveldb/leveldb/iterator"
 	"github.com/syndtr/goleveldb/leveldb/opt"
+	"github.com/syndtr/goleveldb/leveldb/util"
 )
 
-func newOpCache(vrw ValueReadWriter) *opCache {
+var (
+	collectionId = uint32(0)
+	ldb          *opCacheDb
```

Comment:
This change was inspired by my nomdex work. We can have thousands of these streaming collections going simultaneously. Having an ldb instance for each one of them is problematic for at least two reasons: 
1) the number of directories created and potentially leaked in the case of a crash
2) the amount of memory overhead for each ldb instance. 200-300 instances was causing ~10Gb of memory usage that completely disappeared with this change.

I agree about singletons but in this case the ownership issues don't seem to be much more complicated than what we had before.